File: WEEK5_23_2_26/Validations.java
import java.util.regex.Pattern;

public class Validations {
    private static final Pattern EMAIL = Pattern.compile("^[\\w\\.-]+@[\\w\\.-]+\\.\\w+$");
    private static final Pattern PHONE = Pattern.compile("^\\+?[0-9 ()-]{7,20}$");

    // Basic email check (illustrative)
    public static boolean isEmail(String s) {
        if (s == null) return false;
        return EMAIL.matcher(s).matches();
    }

    // Permissive phone check
    public static boolean isPhone(String s) {
        if (s == null) return false;
        return PHONE.matcher(s).matches();
    }

    // True only if the value can be parsed to a positive integer
    public static boolean isPositiveInt(String x) {
        if (x == null) return false;
        try {
            return Integer.parseInt(x) > 0;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    public static void main(String[] args) {
        String[] samples = {"alice@example.com", "bad@.com", "+1 800 555 0199", "12345"};
        for (String s : samples) {
            System.out.println(s + " email? " + isEmail(s) + " phone? " + isPhone(s));
        }

        String[] vals = {"10", "5", "-2", "x"};
        for (String v : vals) System.out.println(v + " positive_int? " + isPositiveInt(v));
    }
}

File: WEEK5_23_2_26/validations.py
"""Validation helpers: email, phone, positive integer.
Small, dependency-free implementations for learning.
"""
import re


def is_email(s):
    # Basic email pattern check (illustrative, not exhaustive)
    if not isinstance(s, str):
        return False
    return re.match(r"^[\w\.-]+@[\w\.-]+\.\w+$", s) is not None


def is_phone(s):
    # Permissive phone format check (allows digits, spaces, parens, hyphens)
    if not isinstance(s, str):
        return False
    return re.match(r"^\+?[0-9 ()-]{7,20}$", s) is not None


def is_positive_int(x):
    # Return True only for values that cast to a positive integer
    try:
        return int(x) > 0
    except (ValueError, TypeError):
        return False


def main():
    # Small demonstration of validation helpers
    samples = ["alice@example.com", "bad@.com", "+1 800 555 0199", "12345"]
    for s in samples:
        print(s, 'email?', is_email(s), 'phone?', is_phone(s))
    for v in [10, '5', -2, 'x']:
        print(v, 'positive_int?', is_positive_int(v))


if __name__ == '__main__':
    main()
